### DISCUSSION TITLE
Not understanding popper invocation for user profile.

### DISCUSSION THREAD

**Discussion by JohnDuerden** [POPWorldMedia/POPForums#306] [General]
Great project - exactly what I need. I have integrated PopForums via Nuget to my project. I am using Azure AD B2C for authentication using a custom identify framework. Oauth is used to authenticate to PopForums. I have used the same _layout for the "main site" and the PopForums "site". Mostly things are working within the _layout (which includes a script reference to popper.js) but invoking the user profile dropdown in PopForums I get "cannot use import from outside a module". The _layout does include jquery for the "main site". I cannot see where popper is included in your git repo.

As a secondary, not important, question I am wondering if all my wiring up to switch between authentication schemes from my "main site" to "PopForums site" could be simplified. I basically signout / signin when switching between the sites and there are numerous places on the pages where that can happen but I think I have caught most of them. It seem you can have multiple identities in my situation and wonder if there is a way to use that functionality that I am not aware of.  You clearly are very skilled at this stuff. Thanks!

**Comment by jeffputz** [comment]
Hi John... are you using one of the releases? Starting with v18, there are no dependencies on jQuery. I've never seen that error, so if you're using the code right off of `main` (the future v19), I'd like to know more. In that next version, all of the front-end stuff is TypeScript, so it does have to be complied.

You can have multiple identities, this is true, and the `User` property in views and controllers and such surfaces the first one in its `Identity` property. For my own sites, I've used the forum's user stuff across the entire app. If you have `app.UsePopForumsAuth()` in your `Program/Startup` then it will be populated on every request unless you have the ` PopForumsAuthorizationIgnore` attribute on the action. That's an extension that adds the middleware to the pipeline that uses the ASP.NET cookie auth to figure out who the user is, verify it's an active account and then put it into that `HttpContext.User`. The user and the profile are also put into the context `Items` collection, which is where the `IUserRetrievalShim` can get at it from everywhere.

So here's the rub... I'm not confident that I've found every instance of where the user is being fetched right out of `HttpContext`, which is the way it worked all over the place back in the day. I wanted instead to rely on the shim, meaning I could just not add the user to context. I should revisit that.

In the mean time, yes, you can see that the login [uses cookie auth with the forum scheme](https://github.com/POPWorldMedia/POPForums/blob/e9eefd1cb9abd97dc2c4704b8b0fc86e15e94fcb/src/PopForums.Mvc/Areas/Forums/Controllers/IdentityController.cs#L192), then [the middleware authenticates on the same scheme](https://github.com/POPWorldMedia/POPForums/blob/e9eefd1cb9abd97dc2c4704b8b0fc86e15e94fcb/src/PopForums.Mvc/Areas/Forums/Authorization/PopForumsAuthorizationMiddleware.cs#L20).

I wouldn't mind hearing more about your use of Azure AD and the OAuth built into the forum! I haven't tried it, but I'm considering some kind of enterprisey option to make that more streamlined, and maybe even update the forum copy of the identity (name, email) as necessary.

**Reply by JohnDuerden** [reply]
Thanks Jeff for the quick response.

I have figured it out. I was pulling in popper in a separate script in the layout. Looking at the Bootstrap 5.2 docs I see they now have a Bootstrap.bundle.js which I think is new? Anyway switching to pulling Bootstrap.bundle.js instead of popper.js  fixed the issue.

I am just using a Nuget pull which I think is a little behind the code on github main branch. I will look into pulling it off the Github main branch to get more current. The other "site" does use jQuery. The other "site" also has its own middleware to add group claims for the user and maybe there is something there I need to change to allow multiple identities.  The other "site" middleware is very similar to PopForums middleware.

This is the action method to switch to the forums:

/// <summary>
        /// Action method which interface to the Forum website to, if necessary, create 
        /// a user. Login to the Forum website is via OAuth2.
        /// </summary>
        /// <returns>Redirects the user to the forum website logging them in</returns>
        [Authorize(Policy = "Member")]
        public RedirectResult PopForum()
        {
            string oauthLink = GetForumOauthLink("~/Forums");
            return Redirect(oauthLink);
        }

and this is the code (based very much on your example - thanks!):

/// <summary>
        /// The method generates a link with an oAuth Json Web Token (JWT) which can be used
        /// to log in to a PopForums url. Refer to:
        /// [External login PopForums](https://popworldmedia.github.io/POPForums/externalloginconfig.html)
        /// and
        /// [Sample external login PopForums app](https://github.com/POPWorldMedia/POPIdentity/tree/main/src/PopIdentity.Sample).
        /// </summary>
        /// <param name="returnUrl">PopForums url to include an oAuth JWT</param>
        /// <returns>A url that can be used in a redirect to a PopForums url</returns>
        private string GetForumOauthLink(string returnUrl)
        {
            // Following logs in to PopForums using [PopIdentity]("https://github.com/POPWorldMedia/POPIdentity")
            string state = _stateHashingService.SetCookieAndReturnHash();
            // This URL has to be specified as legal by whatever provider you're using
            string oauthRedirect = "https://rmraCore.azurewebsites.net/Forums/Identity/CallbackHandler";
            // Use following link for local debugging
            //string oauthRedirect = "https://localhost:44321/Forums/Identity/CallbackHandler";
            OAuth2LoginUrlGenerator linkGenerator = new();
            // choose the scope you're looking for
            List<string> scopes = new(new[] { "openid", "email" });
            // PopIdentity provides a method to generate the url to login to PopForums.
            // Url will include the OAuth2 token generated.
            string oauthLink = linkGenerator.GetUrl(_popIdentitySettings.oAuth2Jwt.LoginUrl,
                                                    _popIdentitySettings.oAuth2Jwt.ClientID,
                                                    oauthRedirect, state, scopes);
            ProviderType providerType = ProviderType.OAuth2;
            ExternalLoginState loginState = new() { ProviderType = providerType, ReturnUrl = returnUrl };
            _externalLoginTempService.Persist(loginState);
            _logger.LogDebug("PopForum oauthlink: {oauthLink}", oauthLink);
            return oauthLink;
        }

I found that if the code signed out of PopForums where appropriate then the main "site" identity appeared so the middleware includes this crude way of handling that (I suspect I have not got all instances of Forum urls)

// Create a new scope
                using var scope = _serviceProvider.CreateScope();
                _adGraphClient = scope.ServiceProvider.GetService<IADGraphClient>();
                _adGraphClient.SetAuthenticationContext(ConfigSettings.TenantId, ConfigSettings.ClientId,
                                                        ConfigSettings.ClientSecret);
                _userManager = scope.ServiceProvider.GetService<UserManager<rmraData.Models.MembershipLists>>();
                AuthenticateResult authResult = await context.AuthenticateAsync(OpenIdConnectDefaults.AuthenticationScheme);
                AuthenticateResult pfauthResult = await context.AuthenticateAsync(PopForumsAuthorizationDefaults.AuthenticationScheme);
                _logger.LogDebug("OpenID authenticate result: {authResult.Succeeded}", authResult.Succeeded);
                var requestPath = context.Request.Path;
                if ((!requestPath.Value.ToUpper().Contains("/FORUM")) && 
                    (!requestPath.Value.ToUpper().Contains("/FORUM/LISTLATESTTOPICS")) &&
                    (!requestPath.Value.ToUpper().Contains("/TOPICSHUB")))
                {
                    await context.SignOutAsync(PopForumsAuthorizationDefaults.AuthenticationScheme);
                }

Thanks again for a great project.
